File: src/nativeTools.ts
import { createHash } from "crypto";
import { Tool } from "@modelcontextprotocol/sdk/types.js";
import { getSystemToolDefinitions } from "./systemTools";
import type {
  ImageContent,
  MessageParam,
  TextContent,
  ToolResultContent,
} from "./types";

export const MAX_TOOL_RESULT_TEXT_CHARACTERS = 100_000;
export const TOOL_RESULT_TRUNCATION_MARKER = "\n...truncated";

export interface NativeToolDefinition {
  apiName: string;
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
}

export type AgentToolEvent =
  | {
      type: "tool_use";
      id: string;
      name: string;
      input: Record<string, unknown>;
      serverTool: boolean;
    }
  | {
      type: "tool_result";
      toolUseId: string;
      name: string;
      content: string;
      isError: boolean;
      serverTool: boolean;
    };

const AGENT_TOOL_EVENT_PREFIX = "\u0000agent_tool:";

export function encodeAgentToolEvent(event: AgentToolEvent): string {
  return AGENT_TOOL_EVENT_PREFIX + JSON.stringify(event);
}

export function isAgentToolEvent(token: string): boolean {
  return token.startsWith(AGENT_TOOL_EVENT_PREFIX);
}

export function decodeAgentToolEvent(
  token: string,
): AgentToolEvent | undefined {
  try {
    return JSON.parse(
      token.substring(AGENT_TOOL_EVENT_PREFIX.length),
    ) as AgentToolEvent;
  } catch {
    return undefined;
  }
}

const VALID_NAME = /^[A-Za-z0-9_-]{1,64}$/;

export function usesNativeTools(modelName: string): boolean {
  return !modelName.toLowerCase().startsWith("google");
}

export function nativeToolName(name: string): string {
  if (VALID_NAME.test(name)) {
    return name;
  }
  const readable =
    name.replace(/[^A-Za-z0-9_-]/g, "_").replace(/^_+|_+$/g, "") || "tool";
  const digest = createHash("sha256")
    .update(name)
    .digest("hex")
    .substring(0, 10);
  return `${readable.substring(0, 53)}_${digest}`;
}

export function buildNativeTools(
  groupedTools: ReadonlyMap<string, ReadonlyMap<string, Tool>>,
): NativeToolDefinition[] {
  const definitions = [...getSystemToolDefinitions()];
  for (const tools of groupedTools.values()) {
    definitions.push(...tools.values());
  }
  return definitions.map((tool) => ({
    apiName: nativeToolName(tool.name),
    name: tool.name,
    description: tool.description || "",
    inputSchema: tool.inputSchema as Record<string, unknown>,
  }));
}

export function canonicalToolName(
  apiName: string,
  tools: readonly NativeToolDefinition[],
): string {
  return tools.find((tool) => tool.apiName === apiName)?.name || apiName;
}

export function apiToolName(
  name: string,
  tools: readonly NativeToolDefinition[],
): string {
  return (
    tools.find((tool) => tool.name === name)?.apiName || nativeToolName(name)
  );
}

export function anthropicToolSchemas(
  tools: readonly NativeToolDefinition[],
): Record<string, unknown>[] {
  return tools.map((tool) => ({
    name: tool.apiName,
    description: tool.description,
    input_schema: tool.inputSchema,
  }));
}

export function openaiChatToolSchemas(
  tools: readonly NativeToolDefinition[],
): Record<string, unknown>[] {
  return tools.map((tool) => ({
    type: "function",
    function: {
      name: tool.apiName,
      description: tool.description,
      parameters: tool.inputSchema,
    },
  }));
}

export function openaiResponsesToolSchemas(
  tools: readonly NativeToolDefinition[],
): Record<string, unknown>[] {
  return tools.map((tool) => ({
    type: "function",
    name: tool.apiName,
    description: tool.description,
    parameters: tool.inputSchema,
  }));
}

function escapeXml(value: string): string {
  return value
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

export function unescapeXml(value: string): string {
  return value
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&amp;/g, "&");
}

function renderParamValue(value: unknown): string {
  const rendered =
    typeof value === "string"
      ? value
      : value === undefined
      ? ""
      : JSON.stringify(value);
  if (!/<\/?cmd:|<!\[CDATA\[|\]\]>/.test(rendered)) return rendered;
  return `<![CDATA[${rendered.replace(/\]\]>/g, "]]]]><![CDATA[>")}]]>`;
}

export function renderToolCall(
  name: string,
  input: Record<string, unknown>,
): string {
  const params = Object.entries(input)
    .map(
      ([key, value]) =>
        `<cmd:param name="${escapeXml(key)}">${renderParamValue(
          value,
        )}</cmd:param>`,
    )
    .join("\n");
  return `\n<cmd:tool_call>\n<cmd:tool_name>${escapeXml(
    name,
  )}</cmd:tool_name>\n${params}${params ? "\n" : ""}</cmd:tool_call>`;
}

export function renderToolCallFromArguments(
  name: string,
  argumentsJson: string,
): string {
  return renderToolCall(name, parseNativeArguments(argumentsJson) ?? {});
}

export function renderServerToolResult(result: string): string {
  return result;
}

export function toolResultText(value: unknown): string {
  if (typeof value === "string") return value;
  if (Array.isArray(value)) {
    return value.map(toolResultText).filter(Boolean).join("\n");
  }
  if (value && typeof value === "object") {
    const result = value as Record<string, unknown>;
    for (const key of [
      "output",
      "aggregatedOutput",
      "stdout",
      "text",
      "content",
      "message",
      "error",
      "structuredContent",
      "structured_content",
    ]) {
      if (result[key] === undefined) continue;
      const text = toolResultText(result[key]);
      if (text) return text;
    }
    return "";
  }
  return value === undefined || value === null ? "" : String(value);
}

function truncatedToolResultRawText(
  rawText: string,
  content: readonly (TextContent | ImageContent)[],
): string {
  const body = content
    .map((part) =>
      part.type === "text" ? part.value : `![Tool result image](${part.path})`,
    )
    .join("\n\n");
  return /<tool_result>[\s\S]*?<\/tool_result>/.test(rawText)
    ? `<tool_result>\n${body}\n</tool_result>`
    : body;
}

function truncateToolResult(result: ToolResultContent): ToolResultContent {
  const textCharacters = result.content.reduce(
    (total, part) => total + (part.type === "text" ? part.value.length : 0),
    0,
  );
  if (textCharacters <= MAX_TOOL_RESULT_TEXT_CHARACTERS) return result;

  let remaining =
    MAX_TOOL_RESULT_TEXT_CHARACTERS - TOOL_RESULT_TRUNCATION_MARKER.length;
  const content: Array<TextContent | ImageContent> = [];
  for (const part of result.content) {
    if (part.type === "image") {
      content.push(part);
      continue;
    }
    if (remaining <= 0) continue;
    const value = part.value.substring(0, remaining);
    remaining -= value.length;
    if (value) content.push({ type: "text", value });
  }
  content.push({ type: "text", value: TOOL_RESULT_TRUNCATION_MARKER });

  return {
    ...result,
    content,
    rawText: truncatedToolResultRawText(result.rawText, content),
  };
}

export function truncateToolResultsForApi(
  messages: readonly MessageParam[],
): MessageParam[] {
  return messages.map((message) => ({
    ...message,
    content: message.content.map((item) =>
      item.type === "tool_result" ? truncateToolResult(item) : item,
    ),
  }));
}

export function assignDeterministicToolIds(
  messages: readonly MessageParam[],
): MessageParam[] {
  const pendingIds: string[] = [];
  let callIndex = 0;
  let orphanResultIndex = 0;

  return messages.map((message) => {
    if (
      message.role === "user" &&
      !message.content.some((item) => item.type === "tool_result")
    ) {
      pendingIds.length = 0;
    }
    return {
      ...message,
      content: message.content.map((item) => {
        if (item.type === "tool_use") {
          const id = `chatmd_call_${callIndex}`;
          callIndex++;
          pendingIds.push(id);
          return { ...item, id };
        }
        if (item.type === "tool_result") {
          let toolUseId = pendingIds.shift();
          if (!toolUseId) {
            toolUseId = `chatmd_orphan_result_${orphanResultIndex}`;
            orphanResultIndex++;
          }
          return { ...item, toolUseId };
        }
        return item;
      }),
    };
  });
}

export function parseServerToolResult(value: string): {
  id?: string;
  result: string;
} {
  const matched =
    /^\s*<cmd:tool_id>([^<]*)<\/cmd:tool_id>[ \t]*(?:\r?\n)?/.exec(value);
  if (!matched) return { result: value };
  return {
    id: unescapeXml(matched[1]),
    result: value.substring(matched[0].length),
  };
}

export function parseNativeArguments(
  value: string,
): Record<string, unknown> | undefined {
  try {
    const parsed: unknown = JSON.parse(unescapeXml(value));
    return parsed !== null &&
      typeof parsed === "object" &&
      !Array.isArray(parsed)
      ? (parsed as Record<string, unknown>)
      : undefined;
  } catch {
    return undefined;
  }
}

export function paramsFromInput(
  input: Record<string, unknown>,
): Record<string, string> {
  return Object.fromEntries(
    Object.entries(input).map(([name, value]) => [
      name,
      typeof value === "string" ? value : JSON.stringify(value),
    ]),
  );
}

export function inputFromParams(
  params: Record<string, string>,
): Record<string, unknown> {
  return Object.fromEntries(
    Object.entries(params).map(([name, value]) => {
      try {
        return [name, JSON.parse(value)];
      } catch {
        return [name, value];
      }
    }),
  );
}

File: python/src/chatmd/providers/native_tools.py
"""Provider-native tool schemas and the persistent chat representation."""

from __future__ import annotations

import hashlib
import json
import re
from collections import deque
from collections.abc import Mapping, Sequence
from dataclasses import dataclass, replace
from typing import Any
from xml.sax.saxutils import escape, unescape

from ..tools.system_tools import get_system_tool_definitions
from ..types import (
    Content,
    ImageContent,
    McpToolDefinition,
    MessageParam,
    TextContent,
    ToolCall,
    ToolResultContent,
    ToolUseContent,
)

_VALID_NAME_RE = re.compile(r"^[A-Za-z0-9_-]{1,64}$")
_INVALID_NAME_RE = re.compile(r"[^A-Za-z0-9_-]")
_SERVER_RESULT_ID_RE = re.compile(r"^\s*<cmd:tool_id>([^<]*)</cmd:tool_id>[ \t]*(?:\r?\n)?")
MAX_TOOL_RESULT_TEXT_CHARACTERS = 100_000
TOOL_RESULT_TRUNCATION_MARKER = "\n...truncated"


@dataclass(frozen=True)
class NativeToolDefinition:
    """One MCP tool plus the API-safe name exposed to a model provider."""

    api_name: str
    name: str
    description: str | None
    input_schema: dict[str, Any]


def uses_native_tools(model_name: str) -> bool:
    return not model_name.lower().startswith("google")


def native_tool_name(name: str) -> str:
    """Return a stable OpenAI/Anthropic-compatible alias for an MCP tool name."""
    if _VALID_NAME_RE.fullmatch(name):
        return name
    readable = _INVALID_NAME_RE.sub("_", name).strip("_") or "tool"
    digest = hashlib.sha256(name.encode("utf-8")).hexdigest()[:10]
    return f"{readable[:53]}_{digest}"


def build_native_tools(
    grouped_tools: Mapping[str, Mapping[str, McpToolDefinition]],
) -> list[NativeToolDefinition]:
    definitions = [*get_system_tool_definitions()]
    definitions.extend(tool for tools in grouped_tools.values() for tool in tools.values())
    return [
        NativeToolDefinition(
            api_name=native_tool_name(tool.name),
            name=tool.name,
            description=tool.description,
            input_schema=tool.input_schema,
        )
        for tool in definitions
    ]


def canonical_tool_name(api_name: str, tools: Sequence[NativeToolDefinition]) -> str:
    return next((tool.name for tool in tools if tool.api_name == api_name), api_name)


def api_tool_name(name: str, tools: Sequence[NativeToolDefinition]) -> str:
    return next((tool.api_name for tool in tools if tool.name == name), native_tool_name(name))


def anthropic_tool_schemas(tools: Sequence[NativeToolDefinition]) -> list[dict[str, Any]]:
    return [
        {
            "name": tool.api_name,
            "description": tool.description or "",
            "input_schema": tool.input_schema,
        }
        for tool in tools
    ]


def openai_chat_tool_schemas(tools: Sequence[NativeToolDefinition]) -> list[dict[str, Any]]:
    return [
        {
            "type": "function",
            "function": {
                "name": tool.api_name,
                "description": tool.description or "",
                "parameters": tool.input_schema,
            },
        }
        for tool in tools
    ]


def openai_responses_tool_schemas(
    tools: Sequence[NativeToolDefinition],
) -> list[dict[str, Any]]:
    return [
        {
            "type": "function",
            "name": tool.api_name,
            "description": tool.description or "",
            "parameters": tool.input_schema,
        }
        for tool in tools
    ]


def render_tool_call_start(name: str) -> str:
    return f"\n<cmd:tool_call>\n<cmd:tool_name>{escape(name)}</cmd:tool_name>\n<cmd:arguments>"


def render_tool_arguments_delta(delta: str) -> str:
    return escape(delta)


def render_tool_call_end() -> str:
    return "</cmd:arguments>\n</cmd:tool_call>"


def render_tool_call(name: str, input_: Mapping[str, Any]) -> str:
    arguments = json.dumps(input_, separators=(",", ":"), ensure_ascii=False)
    return (
        render_tool_call_start(name)
        + render_tool_arguments_delta(arguments)
        + render_tool_call_end()
    )


def render_server_tool_result(result: str) -> str:
    return result


def tool_result_text(value: object) -> str:
    if isinstance(value, str):
        return value
    if isinstance(value, list):
        return "\n".join(filter(None, (tool_result_text(item) for item in value)))
    if isinstance(value, dict):
        for key in (
            "output",
            "aggregatedOutput",
            "stdout",
            "text",
            "content",
            "message",
            "error",
            "structuredContent",
            "structured_content",
        ):
            if key not in value:
                continue
            text = tool_result_text(value[key])
            if text:
                return text
        return ""
    dump = getattr(value, "model_dump", None)
    if callable(dump):
        return tool_result_text(dump(mode="json", by_alias=True, exclude_none=True))
    return "" if value is None else str(value)


def _truncated_tool_result_raw_text(
    raw_text: str, content: Sequence[TextContent | ImageContent]
) -> str:
    body = "\n\n".join(
        part.value if isinstance(part, TextContent) else f"![Tool result image]({part.path})"
        for part in content
    )
    if re.search(r"<tool_result>[\s\S]*?</tool_result>", raw_text):
        return f"<tool_result>\n{body}\n</tool_result>"
    return body


def _truncate_tool_result(result: ToolResultContent) -> ToolResultContent:
    text_characters = sum(
        len(part.value) for part in result.content if isinstance(part, TextContent)
    )
    if text_characters <= MAX_TOOL_RESULT_TEXT_CHARACTERS:
        return result

    remaining = MAX_TOOL_RESULT_TEXT_CHARACTERS - len(TOOL_RESULT_TRUNCATION_MARKER)
    content: list[TextContent | ImageContent] = []
    for part in result.content:
        if isinstance(part, ImageContent):
            content.append(part)
            continue
        if remaining <= 0:
            continue
        value = part.value[:remaining]
        remaining -= len(value)
        if value:
            content.append(TextContent(value=value))
    content.append(TextContent(value=TOOL_RESULT_TRUNCATION_MARKER))

    return replace(
        result,
        content=content,
        raw_text=_truncated_tool_result_raw_text(result.raw_text, content),
    )


def truncate_tool_results_for_api(
    messages: Sequence[MessageParam],
) -> list[MessageParam]:
    return [
        MessageParam(
            role=message.role,
            content=[
                _truncate_tool_result(item) if isinstance(item, ToolResultContent) else item
                for item in message.content
            ],
        )
        for message in messages
    ]


def assign_deterministic_tool_ids(
    messages: Sequence[MessageParam],
) -> list[MessageParam]:
    pending_ids: deque[str] = deque()
    call_index = 0
    orphan_result_index = 0
    normalized: list[MessageParam] = []

    for message in messages:
        if message.role == "user" and not any(
            isinstance(item, ToolResultContent) for item in message.content
        ):
            pending_ids.clear()
        content: list[Content] = []
        for item in message.content:
            if isinstance(item, ToolUseContent):
                call_id = f"chatmd_call_{call_index}"
                call_index += 1
                pending_ids.append(call_id)
                content.append(replace(item, id=call_id))
            elif isinstance(item, ToolResultContent):
                if pending_ids:
                    call_id = pending_ids.popleft()
                else:
                    call_id = f"chatmd_orphan_result_{orphan_result_index}"
                    orphan_result_index += 1
                content.append(replace(item, tool_use_id=call_id))
            else:
                content.append(item)
        normalized.append(MessageParam(role=message.role, content=content))
    return normalized


def parse_server_tool_result(value: str) -> tuple[str | None, str]:
    """Strip the legacy ID prefix while accepting current ID-free content."""
    matched = _SERVER_RESULT_ID_RE.match(value)
    if matched is None:
        return None, value
    return unescape(matched.group(1)), value[matched.end() :]


def decode_tool_arguments(value: str) -> dict[str, Any] | None:
    try:
        parsed = json.loads(unescape(value))
    except (json.JSONDecodeError, TypeError):
        return None
    return parsed if isinstance(parsed, dict) else None


def params_from_input(value: Mapping[str, Any]) -> dict[str, str]:
    return {
        key: item if isinstance(item, str) else json.dumps(item, separators=(",", ":"))
        for key, item in value.items()
    }


def input_from_params(value: Mapping[str, str]) -> dict[str, Any]:
    output: dict[str, Any] = {}
    for key, item in value.items():
        try:
            output[key] = json.loads(item)
        except json.JSONDecodeError:
            output[key] = item
    return output


def tool_call_input(call: ToolCall) -> dict[str, Any]:
    return call.input if call.input is not None else input_from_params(call.params)
